server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -171,6 +171,7 @@
     private final DiffableStringMap hashesOfConsistentSettings;
     private final ImmutableOpenMap<String, IndexMetaData> indices;
     private final ImmutableOpenMap<String, IndexTemplateMetaData> templates;
+    private final ImmutableOpenMap<String, ComponentTemplate> componentTemplates;
```

Comment:
I think that component templates should be stored as custom metadata (like stored scripts or ingest pipelines). This a lot of the repetitive serialization / diffable code can be be removed.

Reply:
Hmm... I don't think that we have too much serialization code, but I think I'd rather treat component templates as a First Class object, and store it right next to the regular templates (rather than have templates be part of the metadata and component templates be part of "custom" metadata)

@martijnvg do you feel strongly about them needing to be Custom metadata instead of stored next to the regular templates?

Reply:
If component templates implement `MetaData.Custom` then no changes are required to this file and the `DiffableUtils.java ` file. I have a strong preference to implement component templates as custom metadata.

I think we can add a getter to the `MetaData` class and a setter to the `MetaData.Builder` class (next to the methods for index templates) in order to advertise component templates as first class concept in cluster metadata.

Ingest pipelines and stored stored scripts are first class concepts too and these are stored as custom metadata too. I think if index templates were coded today then I would have the same strong preference as well.

Reply:
That sounds good to me then, good point about just making the getter first-class. I'll change it to be `Custom`-based

Reply:
Okay, this required about 200 more lines of code, but switched this to use `MetaData.Custom`